httpclient/src/main/kotlin/tbdex/sdk/httpclient/Utils.kt

```diff
@@ -33,35 +37,93 @@ fun getPfiServiceEndpoint(pfiDid: String): String {
+  val assertionMethod: VerificationMethod = didResolutionResult.didDocument.findAssertionMethodById(assertionMethodId)
 
   // TODO: ensure that publicKeyJwk is not null
   val publicKeyJwk = JWK.parse(assertionMethod.publicKeyJwk)
   val keyAlias = did.keyManager.getDeterministicAlias(publicKeyJwk)
 
+  // TODO: figure out how to make more reliable since algorithm is technically not a required property of a JWK
   val algorithm = publicKeyJwk.algorithm
   val jwsAlgorithm = JWSAlgorithm.parse(algorithm.toString())
 
-  val jwsHeader = JWSHeader.Builder(jwsAlgorithm)
-    .keyID(assertionMethod.id.toString())
-    .build()
+  val kid = when (assertionMethod.id.isAbsolute) {
+    true -> assertionMethod.id.toString()
+    false -> "${did.uri}${assertionMethod.id}"
+  }
 
-  val payload = mapOf("timestamp" to Instant.now().toString())
-  val jwsPayload = Payload(payload)
-  val base64UrlEncodedPayload = jwsPayload.toBase64URL().toString()
+  val jwtHeader = JWSHeader.Builder(jwsAlgorithm)
+    .type(JOSEObjectType.JWT)
+    .keyID(kid)
+    .build()
 
-  val jwsObject = JWSObject(jwsHeader, jwsPayload)
-  val toSign = jwsObject.signingInput
+  val now = Instant.now()
+  val exp = now.plusSeconds(60)
+  val jwtPayload = JWTClaimsSet.Builder()
+    .audience(pfiDid)
+    .issuer(did.uri)
+    .expirationTime(Date.from(exp))
+    .issueTime(Date.from(now))
+    .jwtID(UUID.randomUUID().toString())
+    .build()
 
-  val signedBytes = did.keyManager.sign(keyAlias, toSign)
-  val base64UrlEncodedSignature = Base64URL(Convert(signedBytes).toBase64Url(padding = false))
-  val base64UrlEncodedHeader = jwsHeader.toBase64URL()
+  val jwtObject = SignedJWT(jwtHeader, jwtPayload)
+  val toSign = jwtObject.signingInput
+  val signatureBytes = did.keyManager.sign(keyAlias, toSign)
 
+  val base64UrlEncodedHeader = jwtHeader.toBase64URL()
+  val base64UrlEncodedPayload = jwtPayload.toPayload().toBase64URL()
+  val base64UrlEncodedSignature = Base64URL(Convert(signatureBytes).toBase64Url(padding = false))
 
   return "$base64UrlEncodedHeader.$base64UrlEncodedPayload.$base64UrlEncodedSignature"
+}
+
+/**
+ * Verify request token
+ *
+ * @param token JWT bearer token received from the requester
+ * @param pfiDid DID of the PFI
+ * @return DID of the requester/JWT token issuer
+ */
+fun verifyRequestToken(token: String, pfiDid: String): String {
+  val claimsSet: JWTClaimsSet
+  try {
+    claimsSet = SignedJWT.parse(token).jwtClaimsSet
+    // todo: resolving header.kid against a didresolver
+    // todo: getting the verificationMethod and publicKeyJwk and algorithmId
+    // todo: checking if signature is valid `signer.verify({...})`
+  } catch (e: Exception) {
+    throw RequestTokenVerificationException(e, "Failed to parse request token")
+  }
+
+  val issuer = claimsSet.issuer
+  val audience = claimsSet.audience
+  val expirationTime = claimsSet.expirationTime
+
+  val requiredKeys = listOf("aud", "iss", "exp", "jti", "iat")
+  requiredKeys.forEach { key ->
```

Comment:
Since this is checking every key, we can use `all`. i.e

```
  require(requiredKeys.all { key ->
    claimsSet.claims.containsKey(key)
  }) {
    throw RequestTokenMissingClaimsException("Missing required claims.")
  }
  ```

or similar. If we want to log out which specific claim is missing, can keep as `forEach` and `require` inside the loop